### ISSUE TITLE
Required some argument in dataloader for setting randomstate

### ISSUE THREAD

**Issue by deeppatel4557** [pytorch/pytorch#39572]
## Motivation
I am really confused regarding how to set random state in dataloader in PyTorch. I had searched PyTorch forums as well as all related issues but nothing works for me. If there is a simple argument regarding the random state of dataloader then it will be very helpful.

## Pitch

I am working on a project in which I need to load image data from directories. All worked fine but as soon as I executed tried to execute the dataloader again then the results were different. I set shuffle to True but I want some random state in dataloader argument similar to that of train_test_split function in scikit-learn library
## Additional context

cc @SsnL

**Comment by SsnL** [comment]
The random sampler created when `shuffle=True` consumes the default RNG. If you need a fixed order, you might as well not use it...

**Comment by deeppatel4557** [comment]
I need the images to be shuffled in random order but I want the order of the randomness to be reproduced when I re-run the dataloader. If I don't apply randomness in order then it hurts the performance I just want to recover that random order when I run the code again. So is there any way of creating a fixed random order at all the time I execute the code while defining shuffle=True.

**Comment by SsnL** [comment]
The RNG consumed by [`RandomSampler`](https://github.com/pytorch/pytorch/blob/479b04e26a59d20b72ad5fdaec819caaad49af75/torch/utils/data/sampler.py#L68) (created with `shuffle=True`) is the default RNG. So seeding the default RNG will make the same code return same ordering.

If you want the ordering the be the same with different code (really code that consumes different amount of RNG), for now you can create a custom `RandomSampler` class that uses a given RNG state. But I agree that we should really add a `generator=` kwarg for both `DataLoader` and all random samplers.

**Comment by antoinebrl** [comment]
The documentation [has a page on reproducibility](https://pytorch.org/docs/stable/notes/randomness.html).
Even when following these good practices, the data loader might not be fully deterministic if multi-threading is used. Mulit-threading makes the randomness "environment dependent" (interrupts, I/O, context swapping, etc)
